core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -834,7 +840,15 @@ public ShareAcquiredRecords acquire(
                 boolean fullMatch = checkForFullMatch(inFlightBatch, firstBatch.baseOffset(), lastOffsetToAcquire);
                 int numRecordsRemaining = maxRecordsToAcquire - acquiredCount;
                 boolean recordLimitSubsetMatch = isRecordLimitMode && checkForRecordLimitSubsetMatch(inFlightBatch, maxRecordsToAcquire, acquiredCount);
-                if (!fullMatch || inFlightBatch.offsetState() != null || recordLimitSubsetMatch) {
+                boolean throttleRecordsDelivery = shouldThrottleRecordsDelivery(inFlightBatch);
+                // Stop acquiring more records if records delivery has to be throttled. The throttling prevents
+                // complete batch to be archived in case of a single record being corrupt.
```

Comment:
```suggestion
                // Stop acquiring more records if records delivery has to be throttled. The throttling prevents
                // complete batch to be archived in case of a single record being corrupt.
                // Below check isolates the current batch/offsets to be delivered individually in subsequent fetches.
```